docs: race conditions between EndpointSlice update and Pod termination (#45593)
cilium-team/cilium: Documentation/network/kubernetes/kubeproxy-free.rst

Replacing these maps with BPF socket storage (:gh-issue:`42658`) would resolve this issue.

Connection drops during service backend termination
***************************************************

In Kubernetes environment, ``Service`` backends are represented by ``EndpointSlice`` objects that (usually) corresponds

> to ``Pods``. When a Pod is deleted (HPA scale-down, rolling update, etc.), two operations are triggered **simultaneously**:

Reviewer: ```suggestion
In Kubernetes environment, ``Service`` backends are represented by ``EndpointSlice`` objects that usually correspond
to ``Pod`` objects. When a Pod is deleted (HPA scale-down, rolling update, etc.), two operations are triggered **simultaneously**:
```

You can mitigate this by defining ``preStop`` in your Pod spec. The following example delays delivery of the ``TERM``
signal to application containers by 2 seconds, giving Cilium more time to react to the EndpointSlice status change:

.. code-block:: yaml

  lifecycle:
    preStop: # Kubernetes 1.29+
      sleep:
        seconds: 2

The required delay depends on the type of Service being terminated. For ``LoadBalancer`` Services, you may need a
higher value (for example, 10 seconds) to give the external load balancer enough time to stop sending traffic to the
node. For ``ClusterIP`` Services, a lower value (for example, 1 second) may be sufficient. Experiment with different
values to find the best setting for your application.

You can read more here:

    * `Pod Termination Flow <https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination-flow>`__: Item 3 mentions the concurrency issue.
    * `Termination process for Pods and their endpoints <https://kubernetes.io/docs/tutorials/services/pods-and-endpoint-termination-flow/>`__: Suggests using ``preStop`` for graceful termination.

.. admonition:: Remember
  :class: attention

>   ``preStop`` must be a lower number than ``terminationGracePeriodSeconds``, because it's time is counted within this

Reviewer: ```suggestion
  ``preStop`` must be a lower number than ``terminationGracePeriodSeconds``, because its time is counted within this
```

Replacing these maps with BPF socket storage (:gh-issue:`42658`) would resolve this issue.

> Connection Drops Caused by Latency Between Cilium Agent and EndpointSlice Status Changes

Reviewer 2: Maybe prefer a high level heading here?

```suggestion
Connection drops during service backend termination
```

Author: I used the same level as the item items in this “Known Issues” area. I’ll simplify the title as you suggested!

****************************************************************************************

> Imagine the following setup::

Reviewer 2: Maybe prefer this?

```suggestion
In Kubernetes environment, ``Service`` backends are represented by ``EndpointSlice`` objects that corresponds to ``Pods``. When a Pod is deleted (HPA scale-down, rolling update, etc.), two operations are triggered simultaneously:

- The ``EndpointSlice`` endpoint is marked as ``Terminating``, instructing CNIs to stop sending traffic to it.
- Kubelet sends a ``SIGTERM`` to the pod's containers.```

Author: Good one, but the backends can be other things like databases, I’ll try to incorporate the suggestion keeping that in mind.

 {client} -> [Service] -> [EndpointSlice] -> [Pod]

When a Pod is deleted (HPA scale-down, rolling update, etc.), two operations are triggered at the same time:

- The ``EndpointSlice`` endpoint is updated to ``Terminating``, instructing CNIs to stop sending traffic to it.
- Kubelet sends a ``SIGTERM`` to the pod's containers.

Because of that, the application may start refusing connections before Cilium can stop directing traffic
to the Pod. This can lead to dropped connections and client errors. As of Kubernetes 1.36, Kubelet can't coordinate with
EndpointSlice updates due to architectural limitations. At the same time, it's not ideal to implement application-side

> workarounds. For example, web application frameworks **should** immediately refuse connections when they receive

Reviewer 2: Not sure if I understand the implication/intent of this example line?

Author: 2 things: avoiding people to think that this is a cilium bug, and also preventing people from thinking that these are their application bug. Its a mix of clarity and peace of mind.

Reviewer: The example is arguably outside the scope and context of using Cilium for this use case, but I could see the extra context here helping someone who is facing this issue to reason about the solution space. Maybe this helps them to understand a better solution without implementing application-side changes.

> workarounds are usually not ideal. For example, it actually does make sense for web application frameworks to immediately

Reviewer: nit to simplify language:

```suggestion
When these happen at the same time, the application may start refusing connections before Cilium can stop directing traffic
to the Pod. This can lead to dropped connections and client errors. As of Kubernetes 1.36, Kubelet can't coordinate with EndpointSlice updates due to architectural limitations. At the same time, it's not ideal to implement application-side
workarounds. For example, web application frameworks **should** immediately
```

Author: I'll keep the `because`, because (pun intended) there's already a `When` in the beginning of the paragraph, and also indicates the direct consequence. The other changes are great.

refuse connections when they receive a ``TERM`` signal, given that semantically, the application shouldn't know that
it's running in a Kubernetes environment.

> docs: race conditions between EndpointSlice update and Pod termination

Reviewer: Looks like a commit title may have unintentionally leaked into the doc content?
```suggestion

```

Author: My bad. And I read this so many times. 🤦‍♂️

A workaround is to add the following to your Pod spec. This delays delivery of the ``TERM`` signal to application

> containers by 2 seconds, giving Cilium more time to react to the EndpointSlice status change:

Reviewer: nit for tone and simpler wording for future readers:
```suggestion
You can mitigate this by defining `preStop` in your Pod spec. The following example delays delivery of the ``TERM`` signal to application
containers by 2 seconds, giving Cilium more time to react to the EndpointSlice status change:
```

.. code-block:: yaml

  lifecycle:
    preStop: # Kubernetes 1.29+
      sleep:
        seconds: 2

The required delay depends on the type of Service being terminated. For ``LoadBalancer`` Services, you may need a
higher value (for example, 10 seconds) to give the external load balancer enough time to stop sending traffic to the
node. For ``ClusterIP`` Services, a lower value (for example, 1 second) may be sufficient. Experiment with different
values to find the best setting for your application.

You can read more here:

    * `Pod Termination Flow <https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination-flow>`__: Item 3 mentions the concurrency issue.
    * `Termination process for Pods and their endpoints <https://kubernetes.io/docs/tutorials/services/pods-and-endpoint-termination-flow/>`__: Suggests using ``preStop`` for graceful termination.

.. admonition:: Remember
  :class: attention

  ``preStop`` uses time from ``terminationGracePeriodSeconds``, so adjust it accordingly to avoid pods being killed

>   before ``preStop`` finishes.

Reviewer: Does this imply that we should recommend users to make sure `terminationGracePeriodSeconds` is greater than the `preStop.sleep` time? I'm reading that between the lines in this admonition. If that's the case, maybe it's worth explicitly stating that in a short sentence.

Author: I rephrased it to be clear. It's also in the doc attached in the reads I suggested above (part of the termination flow).

Replacing these maps with BPF socket storage (:gh-issue:`42658`) would resolve this issue.

Connection Drops Caused by Latency Between Cilium Agent and EndpointSlice Status Changes
****************************************************************************************

Imagine the following setup:

.. code-block:: text

>  {client} -> [Service] -> [EndpointSlice] -> [Pod]

Reviewer 3: Nitpick: Rather than `.. code-block:: text`, we can just create a simple, non-code literal block with the `::` directive, or its shortcut, the ending `::`, like this:

```suggestion
Imagine the following setup::

    {client} -> [Service] -> [EndpointSlice] -> [Pod]
```

Author: I wasn't aware of this trick. Done! ✅
